### ISSUE TITLE
Tasks run after diff when updating

### ISSUE THREAD

**Issue by pawamoy** [copier-org/copier#196]
I'm using tasks to do some clean up after generating a project.

I noticed that updating the project will compare the new version *pre-tasks* with the current version *post-tasks*.

Shouldn't we compare both versions *post-tasks*? i.e. run tasks in the temporary directory (the new version), and then only compare both projects?

Each time I update my project, files that were previously deleted by a task get re-created during the diff, and then re-deleted by the same task.

I also symlink some files using another task, and when updating, symlinking fails because the links already exist. So I have to handle that in my task. It's not so complicated to do (my tasks are written in a single Python script, so I can easily silence exceptions), but it *could* become difficult to handle with more complex tasks and migrations.

What do you think :slightly_smiling_face:?

Maybe users should have control over *when* the task is executed. My use-case would ask for pre-diff and post-diff tasks actually :smile: 

**EDIT:** I think I can implement that with an initial migration. Will try and report back!
**UPDATE:** hmm nope doesn't work. When first generating the project, migrations do not run, so files are not symlinked and the task fails.

Question: what's the difference between `before` and `after` in a migration?
**UPDATE**: 
- get diff
- run migrations `before`
- apply diff
- run migrations `after`

**Comment by pawamoy** [comment]
So I'm lost at the order of execution of tasks and migrations haha

**Comment by pawamoy** [comment]
I guess what I need is an initial migration. Something that runs only once, after generating a project the first time.

**Comment by Yajo** [comment]
> Shouldn't we compare both versions _post-tasks_? i.e. run tasks in the temporary directory (the new version), and then only compare both projects?

Makes sense.



> Maybe users should have control over _when_ the task is executed. My use-case would ask for pre-diff and post-diff tasks actually

Indeed this would be the best.

**Comment by Yajo** [comment]
Possibly also easier to fix after #110.

**Comment by Yajo** [comment]
> Shouldn't we compare both versions _post-tasks_?

As seen in https://github.com/copier-org/copier/issues/239#issuecomment-677426694 it seems this is how it's really working, so I'm closing here. Please reopen later if you find the reproducible bug :blush:
